pkg/webhook/validator_test.go

```diff
@@ -1844,3 +1847,720 @@ func TestValidatePoliciesCancelled(t *testing.T) {
+	}))
+	t.Cleanup(fulcioServer.Close)
+	fulcioURL, err := apis.ParseURL(fulcioServer.URL)
+	if err != nil {
+		t.Fatalf("Failed to parse fake Fulcio URL")
+	}
+
+	rekorServer := httptest.NewServer(http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
+		rw.Write([]byte(rekorResponse))
+	}))
+	t.Cleanup(rekorServer.Close)
+	rekorURL, err := apis.ParseURL(rekorServer.URL)
+	if err != nil {
+		t.Fatalf("Failed to parse fake Rekor URL")
+	}
+
+	var authorityKeyCosignPub *ecdsa.PublicKey
+
+	pems := parsePems([]byte(authorityKeyCosignPubString))
+	if len(pems) > 0 {
+		key, _ := x509.ParsePKIXPublicKey(pems[0].Bytes)
+		authorityKeyCosignPub = key.(*ecdsa.PublicKey)
+	} else {
+		t.Errorf("Error parsing authority key from string")
+	}
+
+	si.Informer().GetIndexer().Add(&corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: system.Namespace(),
+			Name:      secretName,
+		},
+		Data: map[string][]byte{
+			// Random public key (cosign generate-key-pair) 2021-09-25
+			"cosign.pub": []byte(`-----BEGIN PUBLIC KEY-----
+MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEapTW568kniCbL0OXBFIhuhOboeox
+UoJou2P8sbDxpLiE/v3yLw1/jyOrCPWYHWFXnyyeGlkgSVefG54tNoK7Uw==
+-----END PUBLIC KEY-----
+`),
+		},
+	})
+
+	kc := fakekube.Get(ctx)
+	// Setup service acc and fakeSignaturePullSecrets for "default", "cosign-system" and "my-secure-ns" namespace
+	for _, ns := range []string{"default", system.Namespace(), "my-secure-ns"} {
+		kc.CoreV1().ServiceAccounts(ns).Create(ctx, &corev1.ServiceAccount{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "default",
+			},
+		}, metav1.CreateOptions{})
+
+		kc.CoreV1().Secrets(ns).Create(ctx, &corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: "fakeSignaturePullSecrets",
+			},
+			Data: map[string][]byte{
+				"dockerconfigjson": []byte(`{"auths":{"https://index.docker.io/v1/":{"username":"username","password":"password","auth":"dXNlcm5hbWU6cGFzc3dvcmQ="}}`),
+			},
+		}, metav1.CreateOptions{})
+	}
+
+	// Create fake secret in a non-default namespace and patch the default service acc
+	kc.CoreV1().Secrets("my-secure-ns").Create(ctx, &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "fakeSignaturePullSecretsNonDefault",
+		},
+		Data: map[string][]byte{
+			"dockerconfigjson": []byte(`{"auths":{"https://index.docker.io/v1/":{"username":"username","password":"password","auth":"dXNlcm5hbWU6cGFzc3dvcmQ="}}`),
+		},
+	}, metav1.CreateOptions{})
+
+	mergePatch := map[string]interface{}{
+		"imagePullSecrets": map[string]interface{}{
+			"name": "fakeSignaturePullSecretsNonDefault",
+		},
+	}
+	patch, err := json.Marshal(mergePatch)
+	if err != nil {
+		t.Fatalf("Failed to marshal merge patch: %v", err)
+	}
+	kc.CoreV1().ServiceAccounts("my-secure-ns").Patch(ctx, "default", types.MergePatchType, patch, metav1.PatchOptions{})
+
+	v := NewValidator(ctx, secretName)
+
+	cvs := cosignVerifySignatures
+	defer func() {
+		cosignVerifySignatures = cvs
+	}()
+	// Let's just say that everything is verified.
```

Comment:
I'm curious if this actually tests that we use the imagepullsecrets from the correct namespace because we use these stubbed out ones. Or if it was modified it's hard to see because this looks like cut&paste from the above tests. Would it maybe be enough to override one of these 'pass' or 'fail' and check in them that the checkopts have indeed the correct secret referenced in them.

Reply:
I am not sure if we can test it there because most of the errors I saw were raised when trying to call `k8schain.New` while passing `k8schain.Options` without the proper namespace.